enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,411 @@
+browser.  It can be specified, but support may vary by implementation.
+
+## Proposal
+
+Each goal is addressed in the subsections below.
+
+### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is regarded as an HSTS host
+  - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+  - `max-age` decrements over time
+- `includeSubDomains`: if present, applies to any subdomains of the host's domain name
+- `preload`: if present, tells the client to include this host in its host preload list so that
+it never needs to do an initial load to get the HSTS header (note that this is not defined in RFC 6797
+and is therefore client implementation-dependent)
+
+HSTS is currently implemented as a pattern in the HAProxy template and applied to `edge` and
+`reencrypt` routes that match the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+
+#### Enable HSTS for all routes globally, based on the cluster's ingress domain
+To add the header to all HTTP responses from an `edge` or `reencrypt` route, add a new router environment
+variable called `ROUTER_DEFAULT_HSTS`.
+
+Add a new type `IngressControllerHSTSHeaderPolicy` to the API in
+`openshift/api/blob/master/operator/v1/types_ingress.go`, to capture the configuration of the HSTS Policy.
+```go
+type IngressControllerHSTSHeaderPolicy {
+    // max-age is the delta time range in seconds during which the host is regarded as an HSTS host
+	// If set to 0, it negates the effect and the host is no longer regarded as an HSTS host.
+	// max-age` decrements over time.
+    // +required
+    MaxAge             string `json:"max-age"`
+
+    // scope indicates whether this header policy is enabled on a per-router basis (i.e. some
+    // routers will use HSTS and some will not) or it is global across every router in the cluster
+    // ingress domain.  If per-router HSTS configuration scope is desired, set the scope field to "limited".
+    // If global HSTS is desired, set the scope field to "all".  The default is "all".
```

Comment:
Should there be an "off" value?  Hm, I guess `max-age: 0` is the "off" value?  I think it would be clearer to make `scope` a tri-state with off, limited, and all values.

Reply:
Let me think about that a bit.